README.md
# smart-garage-backend

This is the source code for the backend server of the garage. It includes:

* garage-control-server: Controls the opening and closing the garage door
* monitoring-server: This includes:
    - /sensorchange - Detects the open and close of the garage door and writes it to a log file
    - /openalert - Sends an AWS SNS notification if the garage has been open for a long period of time
    - /lastactivity - Gets the last activity of the garage door use when the smart-garage frontend is accessed. The last activity is holds the current status of the garage door.
    - /history - Gets usage log history of the garage door
    - /stream - Redis stream to send push notifications of when garage door is opened, closed, or has been opened for a long period
    - /climate - Gets the current temperature and humidity of the garage
* sensor-server: Door sensor loop which monitors the open/close status of the garage door

Install the prerequisites:

`python3 -m pip install -r requirements.txt`

Run server:

`./autostart.sh`

**Please note:** You will need to install tmux and tmuxinator

To startup automatically on boot add this to /etc/rc.local

`runuser -l pi /home/pi/smart-garage-backend/autostart.sh &`



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2021-03-11

### Added

- Added Garage door control server that toggles the garage door open and closed
- Added Logging Server that logs all usages of the garage to json file
- Added monitoring server to keep track of the current state of the garage door
- Added low level door sensor program to keep track of the door sensor switch
- Switch to Gitflow style development model with this release

## [0.1.1] - 2021-03-11

### Fixed

- Added CORS header: Access-Control-Allow-Origin to allow requests from originating
from the [smart-garage frontend app](https://github.com/husain3/smart-garage)

## [0.2.0] - 2021-03-11

### Added

- SG-13: Ignore json logging file and stop tracking
- SG-14: Change redis message to json format

## [0.2.1] - 2021-03-11

### Fixed

- SG-22: Create json if it doesn't exist and add json placeholders to HTTP responses if no logs exist.

## [0.3.0] - 2021-03-12

### Added

- SG-14: Create autostart script using tmuxinator to allow all servers to be booting in a tmux pane on boot.

## [0.4.0] - 2021-03-14

### Added

- SG-32: Add authentication for garage door toggle

## [0.5.0] - 2021-03-14

### Added

- SG-27: Add browser push notifications every 10 minutes if garage door remains open

## [1.0.0] - 2021-04-05

### Added

- SG-33: Keep attempting to send sensor status on fail
- SG-52: Move logging functionality to monitoring server
- SG-31: Add call to AWS Lambda smart-garage-notifications
- SG-52: Speed up overall performance by removing redis
- SG-48: Add temperature and humidity sensor functionality
- Remove logging-server codebase

sensor-server/sensor.py
import RPi.GPIO as GPIO
from datetime import datetime
from time import sleep
import requests

GPIO.setmode(GPIO.BOARD)
GPIO.setwarnings(False)

GPIO.setup(16, GPIO.IN, pull_up_down=GPIO.PUD_UP)

def door_open():
	print("Door is open")
	time_opened = datetime.now()
	print(time_opened)
	while True:
		try:
			#POST open status date/time to monitoring server
			sensorchange = requests.post('http://192.168.1.104:5001/sensorchange',
										params={'door_status': "OPENED",
												'date': time_opened.strftime("%B %d, %Y"),
                                                'time': time_opened.strftime("%I:%M:%S %p")})
		except Exception as e:
			print(f"CANNOT SEND DOOR OPENED STATUS TO MONITORING SERVER. {e}")
			sleep(2)
			continue
		break
	while(1):
		time_now = datetime.now()
		if((int((time_now-time_opened).total_seconds()) % 600) == 0 and
			(int((time_now-time_opened).total_seconds()) != 0)):
			try:
				#Send GET to monitoring server /openalert endpoint if door is open for 10 minutes
				openalert = requests.post('http://192.168.1.104:5001/openalert',
										 params={'minutes': int((time_now-time_opened).total_seconds() / 60)})
			except Exception as e:
				print(f"CANNOT SEND ALERT TO MONITORING SERVER. {e}")
			print(f"WARNING: Garage has been open for {int((time_now-time_opened).total_seconds() / 60)} minutes")
			sleep(1)
		if(GPIO.input(16) == GPIO.HIGH):
			print(f"Door was opened for {(time_now-time_opened).total_seconds()}")
			return


def door_closed():
	print("Door is closed")
	time_closed = datetime.now()
	print(time_closed)
	while True:
		try:
			#POST close status date/time to monitoring server
			sensorchange = requests.post('http://192.168.1.104:5001/sensorchange',
									params={'door_status': "CLOSED",
                                            'date': time_closed.strftime("%B %d, %Y"),
                                            'time': time_closed.strftime("%I:%M:%S %p")})
		except Exception as e:
			print(f"CANNOT SEND DOOR CLOSED STATUS TO MONITORING SERVER. {e}")
			sleep(2)
			continue
		break
	while(1):
		if(GPIO.input(16) == GPIO.LOW):
			print(f"Door was opened for {(datetime.now()-time_closed).total_seconds()}")
			return

def main():
	try:
		while(1):
			if(GPIO.input(16) == GPIO.LOW):
				door_open()
			else:
				door_closed()
	except KeyboardInterrupt:
		print("Ctrl-C Recieved. Closing sensor.py")
	except Exception as e:
		print(f"Sensor service failed: {e}")




if __name__ == "__main__":
    main()


logging-server/garage-logging.py
from flask import Flask, render_template, request, jsonify, Response
import json
from os import path

app = Flask(__name__)

def write_json(data, filename='garage_logs.json'): 
	with open(filename,'w') as f: 
		json.dump(data, f, indent=4)

@app.route('/garageactivity', methods=['POST'])
def log_garage_activity():
	door_status = request.args.get('door_status')
	date = request.args.get('date')
	time = request.args.get('time')

	if not path.exists("garage_logs.json"):
		initLogJson = '''
{
	"garagedoor_usage_log": [

	]
}
'''

		f = open('garage_logs.json', 'w')
		f.write(initLogJson)
		f.close()

	try:
		if not door_status or not date or not time:
			params_missing = "\n"
			if not door_status:
				params_missing+=str("door_status\n")
			if not date:
				params_missing+=str("date\n")
			if not time:
				params_missing+=str("time\n")

			#TODO: Return missing param arguments as json
			raise AssertionError(f"The following params are missing: {params_missing}")

		with open('garage_logs.json') as json_file:
			data = json.load(json_file)
			log_entries = data["garagedoor_usage_log"]

			# Garage usage log entry to be appended
			log_entry = {"door_status": door_status,
						 "date": date,
						 "time": time
						}

			#Appending log entry given from params to json variable
			log_entries.append(log_entry)
			
			#Write logs into json file
			write_json(data)
		return Response('OK', status=200, mimetype='text/html')
	except AssertionError as a:
		print(f'/garageactivity. Unable to process. Reason: {e}')
		return Response(f'Unable to process. Reason: {a}', status=400, mimetype='text/html')
	except Exception as e:
		print(f'/garageactivity. Unable to process. Reason: {e}')
		return Response(f'Unable to process. Reason: {e}', status=500, mimetype='text/html')

@app.route('/lastactivity', methods=['GET'])
def get_last_activity():
	try:
		if path.exists("garage_logs.json"):
			with open('garage_logs.json') as json_file:
				data = json.load(json_file)
				log_entries = data["garagedoor_usage_log"]

				try:
					last_entry = log_entries[-1:][0]
				except IndexError as I:
					last_entry = {
						"door_status": "N/A",
						"date": "N/A",
						"time": "N/A"
					}

				response = Response(response=json.dumps(last_entry), status=200, mimetype='application/json')
				response.headers["Access-Control-Allow-Origin"] = "*"
				return response

	except Exception as e:
		print(f'/lastactivity. Unable to process. Reason: {e}')
		return Response(f'Unable to process. Reason: {e}', status=500, mimetype='text/html')

@app.route('/history', methods=['GET'])
def get_log_history():
	try:
		with open('garage_logs.json') as json_file:
			data = json.load(json_file)
			log_entries = data["garagedoor_usage_log"]
			log_entries.reverse()
			# print(type(data))
			return Response(response=json.dumps(data), status=200, mimetype='application/json')
	except Exception as e:
		return Response(f'Unable to process. Reason: {e}', status=500, mimetype='text/html')

if __name__ == "__main__":

	if not path.exists("garage_logs.json"):
		initLogJson = '''
{
	"garagedoor_usage_log": [

	]
}
'''
		f = open('garage_logs.json', 'w')
		f.write(initLogJson)
		f.close()

	app.run(debug=True, host='0.0.0.0', port=5002)


garage-control-server/garage-control.py
#	================================================================================================
#	File	: garage-control.py
#	Purpose	: Controls the opening and closing of the garage door
#	================================================================================================

import json
import time
import RPi.GPIO as GPIO
from flask import Flask, render_template, request, jsonify, Response
from flask_httpauth import HTTPBasicAuth
from flask_cors import CORS, cross_origin
from werkzeug.security import generate_password_hash, check_password_hash

from flask_basicauth import BasicAuth

GPIO.setmode(GPIO.BOARD)
GPIO.setwarnings(False)

GPIO.setup(7, GPIO.OUT)
GPIO.output(7, GPIO.HIGH)

app = Flask(__name__)
cors = CORS(app)

# Gets the username and password saved in a separate json file
f = open('config.json')
data = json.load(f)
user = data["user"]
f.close()

app.config['BASIC_AUTH_USERNAME'] = user['username']
app.config['BASIC_AUTH_PASSWORD'] = user['password']

basic_auth = BasicAuth(app)

#	================================================================================================
#	/openclose - Triggers the garage door the open/close
#	================================================================================================
@app.route('/openclose', methods=['POST'])
@cross_origin()
@basic_auth.required
def Garage():
	try:
		GPIO.output(7, GPIO.LOW)
		time.sleep(1)
		GPIO.output(7, GPIO.HIGH)
		time.sleep(2)

		response = Response('OK', status=200, mimetype='text/html')

		return response
	except Exception as e:
		return Response(f'Unable to process. Reason: {e}', status=500, mimetype='text/html')

if __name__ == "__main__":
	app.run(debug=True, host='0.0.0.0', port=5000, ssl_context=('/home/pi/auth/cert.pem', '/home/pi/auth/key.pem'))


monitoring-server/monitoring.py
#	================================================================================================
#	File	: monitoring.py
#	Purpose	: Monitors the garage door, logs usage, and measures indoor climate
#	================================================================================================

from flask import Flask, render_template, request, jsonify, Response, json
from flask_sse import sse
from apscheduler.schedulers.background import BackgroundScheduler
import requests
import redis
import json
import subprocess
import Adafruit_DHT as dht
import threading
import atexit
from flask_cors import CORS, cross_origin

garage_current_state = {}

POOL_TIME = 60  # Seconds

# Climate dict to save temperature and humidity for fast access
current_climate = {}

# Lock to control access to variable
dataLock = threading.Lock()

# Thermometer thread handler
tempThread = threading.Thread()

#	================================================================================================
#	Function to initialize flask server and start climate thread routine
#	================================================================================================
def create_monitoring_server():
	app = Flask(__name__)

	#Set thermometer data pin
	DHT = 17

	#SIGTERM handler
	def interrupt():
		global tempThread
		tempThread.cancel()

	#Read thermometer and humidity sensor
	def getClimate():
		global current_climate
		global tempThread

		with dataLock:
			try:
				humidity = 0.0
				temperature = 0.0

				for i in range(3):
					humidity, temperature = dht.read_retry(dht.DHT22, DHT)
					print(f"{humidity}  {temperature}")

				temperature = round(temperature, 1)
				humidity = round(humidity, 1)

				current_climate["temperature"] = temperature
				current_climate["humidity"] = humidity

			except Exception as e:
				print(f"CANT GET CLIMATE: {e}")

		# Set the next thread to happen
		tempThread = threading.Timer(POOL_TIME, getClimate, ())
		tempThread.start()

	# Initialize climate reading routine
	def getClimateStart():
		# Initialize temperature thread
		global tempThread

		# Create your thread
		tempThread = threading.Timer(0, getClimate())
		tempThread.start()

	# Initiate
	getClimateStart()
	
	# SIGTERM handler registering
	atexit.register(interrupt)
	return app


app = create_monitoring_server()

#	================================================================================================
#	Writes the usage event to a plain text log file
#	================================================================================================
def write_txt(data, filename='garage_logs.txt'):
	with open(filename,'a') as f:
		f.write(f"{data['door_status'].upper()}\t{data['date']}\t{data['time']}\n")


#	================================================================================================
#	/history - Gets full log history of garage door
#	================================================================================================
@app.route('/history', methods=['GET'])
@cross_origin()
def get_log_history():
	#Need to wrap in try/catch
	try:
		content = ''
		content = content.join(str(n) for n in reversed(open("garage_logs.txt").readlines()))
		return render_template("garage_log.html", content=content)
	except Exception as e:
		response = Response(f'Unable to process. Reason: {e}', status=500, mimetype='text/html')
		response.headers["Access-Control-Allow-Origin"] = "*"

		return response


#	================================================================================================
#	/climate - Gets current temperature and humidity of the garage area
#	================================================================================================
@app.route('/climate', methods=['GET'])
def climate():
	try:

		response = Response(response=json.dumps(current_climate), status=200, mimetype='application/json')
		response.headers["Access-Control-Allow-Origin"] = "*"

		return response
	except Exception as e:
		print(f'/climate. Unable to process. Reason: {e}')

		response = Response(f'Unable to process. Reason: {e}', status=500, mimetype='text/html')
		response.headers["Access-Control-Allow-Origin"] = "*"

		return response

#	================================================================================================
#	/lastactivity - Gets the last activity of the garage door
#	================================================================================================
@app.route('/lastactivity', methods=['GET'])
def get_last_activity():
	try:
		#Fastest way to read the last line of a plain text file (last garage usage)
		line = subprocess.check_output(['tail', '-1', "garage_logs.txt"]).decode("utf-8")
		line = line.split()

		last_activity = {
			"door_status": line[0],
			"date": line[1],
			"time": line[2]
		}
		response = Response(response=json.dumps(last_activity), status=200, mimetype='application/json')
		response.headers["Access-Control-Allow-Origin"] = "*"

		return response
	except Exception as e:
		print(f'/lastactivity. Unable to process. Reason: {e}')

		response = Response(f'Unable to process. Reason: {e}', status=500, mimetype='text/html')
		response.headers["Access-Control-Allow-Origin"] = "*"

		return response

#	================================================================================================
#	/openalert - Sends alert that the garage door as been open for a long period
#	================================================================================================
@app.route('/openalert', methods=['POST'])
def send_alert():
	try:
		minutes_opened = request.args.get('minutes')
		if not minutes_opened:
			raise AssertionError("No 'minutes' param sent")
		print(f"WARNING: Garage door has been opened for {minutes_opened} minutes")

		garage_still_open = {}

		garage_still_open['door_status'] = 'STILL_OPEN'
		garage_still_open['minutes_opened'] = minutes_opened

		#Send Amazon SNS text message to phone for notifications
		with open("/home/pi/auth/api_key.json") as json_file:
			data = json.load(json_file)
			url = data["smart-garage-notification-url"]
			api_key = data["x-api-key"]

		aws_response = requests.post(url=url,
									params={
										'door_status': garage_still_open['door_status'],
										'open_duration': garage_still_open['minutes_opened']
										},
									headers={'x-api-key': api_key}
									)

		return Response('OK', status=200, mimetype='text/html')

	except AssertionError as a:
		return Response(f'Unable to process. Reason: {a}', status=400, mimetype='text/html')
	except Exception as e:
		return Response(f'Unable to process. Reason: {e}', status=500, mimetype='text/html')

#	================================================================================================
#	/sensorchange - Changes the open/close status of the garage door
#	================================================================================================
@app.route('/sensorchange', methods=['POST'])
def door_sensor_change():
	door_status = request.args.get('door_status')
	date = request.args.get('date')
	time = request.args.get('time')

	try:
		if not door_status or not date or not time:
			params_missing = "\n"
			if not door_status:
				params_missing+=str("door_status\n")
			if not date:
				params_missing+=str("date\n")
			if not time:
				params_missing+=str("time\n")

			#TODO: Return missing param arguments as json
			raise AssertionError(f"The following params are missing: {params_missing}")

		garage_current_state['door_status'] = door_status
		garage_current_state['date'] = date
		garage_current_state['time'] = time

		#POST sensor change status date/time to AWS to send SMS notification via SES
		with open("/home/pi/auth/api_key.json") as json_file:
			data = json.load(json_file)
			url = data["smart-garage-notification-url"]
			api_key = data["x-api-key"]

		aws_response = requests.post(url=url,
									params={
										'door_status': garage_current_state['door_status']
										},
									headers={'x-api-key': api_key}
									)

		# Garage usage log entry to be appended
		log_entry = {"door_status": request.args.get('door_status'),
						"date": request.args.get('date'),
						"time": request.args.get('time')
					}

		#Write logs into json file
		write_txt(log_entry)

		return Response('OK', status=200, mimetype='text/html')

	except AssertionError as a:
		return Response(f'Unable to process. Reason: {a}', status=400, mimetype='text/html')
	except Exception as e:
		return Response(f'Unable to process. Reason: {e}', status=500, mimetype='text/html')

if __name__ == "__main__":
	app.run(debug=True, host='0.0.0.0', port=5001)
